impl(bigtable): modern `AsyncRowReader`

dbolduc opened the pull request:
Most of the work for #9174

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9287)
<!-- Reviewable:end -->

coryan reviewed (commented):
A few questions and nits.

It is really hard to get a sense of what the tests do, or how they changed.  The coverage looks good, but that is not a perfect way to assess if tests are good enough.  Any thoughts on that?

dbolduc commented on google/cloud/bigtable/internal/async_row_reader_test.cc:
```diff
@@ -12,1147 +12,916 @@
+                         std::move(mock_b));
 }
 
-TEST_F(TableAsyncReadRowsTest, ReadRowNotFound) {
```
So this test fixture was originally for `Table::AsyncReadRow(s)`. (both singular and plural). Now, the test fixture is solely for unit testing the `bigtable_internal::AsyncRowReader` helper class.

These `ReadRow` tests are for [this logic](https://github.com/googleapis/google-cloud-cpp/blob/fd1d0ebe0e347c3074e8a90db94381863f3ccbbf/google/cloud/bigtable/table.cc#L435-L487), which will move to the `DataConnection`. We will test the logic in `data_connection_impl_test.cc`.

dbolduc commented on google/cloud/bigtable/internal/async_row_reader_test.cc:
```diff
@@ -12,1147 +12,916 @@
+        return stream;
+      });
 
-TEST_F(TableAsyncReadRowsTest, ReadRowError) {
```
Same answer as: https://github.com/googleapis/google-cloud-cpp/pull/9287#discussion_r899718671

dbolduc commented on google/cloud/bigtable/internal/async_row_reader_test.cc:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2019 Google LLC
```
Fixed. Changing the license year was unintentional. Thanks for catching it.

dbolduc commented on google/cloud/bigtable/internal/async_row_reader_test.cc:
```diff
@@ -12,1147 +12,916 @@
-  ExpectRow("r1");
-  promises_from_user_cb_[0].set_value(true);
-  ReadRows(1);
+TEST(AsyncRowReaderTest, ParserEofDoesntFailsOnUnfinishedRowIfRowLimit) {
```
Done.

dbolduc commented on google/cloud/bigtable/internal/async_row_reader_test.cc:
```diff
@@ -12,1147 +12,916 @@
+
+  auto mock = std::make_shared<MockBigtableStub>();
+  EXPECT_CALL(*mock, AsyncReadRows)
+      .WillOnce([](CompletionQueue const&, std::unique_ptr<grpc::ClientContext>,
```
Neat.

dbolduc commented:
> It is really hard to get a sense of what the tests do

All the tests have the same structure, so I added comments for the first one. I also wrote in sentences the expectations we are looking for, for the tests that are less obvious. I hope that is what you mean.

>or how they changed.

They hardly look alike. I think mainly due to using a MockStream vs. running everything on a FakeCompletionQueueImpl.

> The coverage looks good, but that is not a perfect way to assess if tests are good enough. Any thoughts on that?

I looked through the code again. I added two tests for code that we were not exercising. I think what we have is good enough.
